Feature/query params

libuk opened the pull request:
### What have I done
As part of the task for removing 3rd party tracking scripts on previews, I've added the `_tapestry` object to `window.__TAPESTRY_DATA__` which contains the following:

```
{
  "requestData": {
    "path": "/:category/:subcategory?/:slug/:id",
    "url": "/life/heathers-tv-reboot-dropped-paramount-america-school-shootings-films-hollywood/210276",
    "isExact": true,
    "params": {
      "category": "life",
      "slug": "heathers-tv-reboot-dropped-paramount-america-school-shootings-films-hollywood",
      "id": "210276"
    },
    "queryParams": {
      "tapestry_hash": "615e29e009eb95bf1a90bcb219fe6b46a38203d4",
      "p": "210276"
    }
  }
}
```
So we now have access to the `tapestry_hash` query param to check for on the client.

jonsherrard commented:
Well howdy @libuk, welcome to Tapestry Lite! This looks really cool and I was just thinking about having the match object available for something I was thinking about.

Could you explain the test changes in a bit more detail? Why would the structure of the object have changed?

libuk commented on src/server/render/tree-to-html.js:
```diff
@@ -7,14 +7,24 @@ export default async ({
-  const loadableState = await getLoadableState(app)
+  const bootstrapData = {
+    ...componentData,
+    requestData: {
```
Hey @jonsherrard this is where the object structure changes as now we'd be passing in the component data as well as the new `requestData` object. Where `bootstrapData` will eventually become `window.__TAPESTRY_DATA__.appData`.

thomasdigby replied:
I wonder if we should nest the `requestData` within a `tapestry`/`tapestryData`/`__tapestry` key? What would be the most obvious key within the WP data that wouldn't clash with any response from WP and is recognisable as Tapestry related

libuk replied:
I like the idea of a `__tapestry` actually. Maybe modifying the object structure to be:

```
appData: {},
__tapestry: {
  requestData: {}
},
ids: []
```

This would eliminate the need to change the app data object structure and provides a place we can use to store data like this in future?

@thomasdigby @jonsherrard

thomasdigby replied:
Bear in mind that we only pass in `appData` to the component, so we'll need to include `__tapestry` in that key

thomasdigby replied:
Actually, let's try and match WP's schema, `_tapestry` should do it.

thomasdigby replied:
Yep that's a good shout, let's do that 👍

thomasdigby commented on test/server/preview.test.js:
```diff
@@ -7,12 +7,29 @@ import CacheManager from '../../src/server/utilities/cache-manager'
 
-const prepareJson = data =>
-  JSON.stringify(data)
+const createRequestData = request => {
```
Are we using this anywhere?

libuk replied:
That needs deleting! Good catch.

thomasdigby commented on src/client/root.js:
```diff
@@ -19,7 +19,7 @@ const Root = config => {
   )
   log('Matched route', { route, match })
-  return <route.component {...match} {...window.__TAPESTRY_DATA__.appData} />
+  return <route.component {...window.__TAPESTRY_DATA__.appData} />
```
How did the experiment go passing `route.component` in with appData?

libuk commented:
With 348e400 I've added a new `_tapestry` key onto the `__TAPESTRY_DATA__` object so as not to change `appData` object structure. I've changed the tests back to reflect this.

thomasdigby commented on test/server/routing.test.js:
```diff
@@ -26,7 +26,7 @@ describe('Handling custom static routes', () => {
       {
         path: '/static-route/:custom',
-        component: props => <p>Param: {props.params.custom}</p>
+        component: props => <p>Param: {props.requestData.params.custom}</p>
```
Should this be `props._tapestry.requestData.params.custom`?

thomasdigby replied:
Ah right, we're spreading it. I think we should keep the key as `_tapestry` to avoid any conflicts within the WP data

libuk replied:
No, this should stay the same as we're spreading the `_tapestry` object, so on a component level, you'll still be able to access all props from `appData` and `_tapestry` at the top level.

libuk replied:
Ok, I'm fine with changing that.

libuk commented:
@thomasdigby It seems we don't have access to the hash server side, as it's only available on the client: 

> The fragment identifier functions differently to the rest of the URI: its processing is exclusively client-side with no participation from the web server, though the server typically helps to determine the MIME type, and the MIME type determines the processing of fragments. **When an agent (such as a Web browser) requests a web resource from a Web server, the agent sends the URI to the server, but does not send the fragment**

https://en.wikipedia.org/wiki/Fragment_identifier

libuk commented:
I'll go ahead and update changelog and version

libuk commented on package.json:
```diff
@@ -16,8 +16,8 @@
-    "test:server": "NODE_ENV=test LOG_LEVEL=silent mocha ./src/**/*.test.js ./test/server/*.test.js --require test/babel-register --bail",
-    "test:integration": "PORT=4321 NODE_ENV=test LOG_LEVEL=silent mocha ./test/integration/*.test.js --require test/babel-register --bail"
+    "test:server": "NODE_ENV=test LOG_LEVEL=silent mocha ./src/**/*.test.js ./test/server/*.test.js --require test/babel-register",
+    "test:integration": "PORT=4321 NODE_ENV=test LOG_LEVEL=silent mocha ./test/integration/*.test.js --require test/babel-register"
```
I was removing the `--bail` flag locally to see the complete list of failed tests. @jonsherrard @thomasdigby let me know if this is OK.

libuk commented:
I'm not authorised to merge 🙇🏾

jonsherrard commented:
Lovely!
